### ISSUE TITLE
Add toArray1D<type>, toArray2D<type> etc. 

### ISSUE THREAD

**Issue by ozgurakpinar** [DiffSharp/DiffSharp#334]
I have been trying to incorporate diffsharp into an existing codebase which is based on `Math.NET Numerics` but I am having trouble with getting standard `vectors` and `matrices` from diffsharp `tensors`.  My intention is to check correctness of current calculations with diffsharp.

```
let t1 = dsharp.tensor [ 0.0 .. 0.2 .. 1.0 ]
let a = t1.toArray()
```
This returns `System.Array` which is incompatible with `'a [ ]`for some reason.

Is there a way to get typed F# arrays (or 2d arrays) from tensors? A good example to what I need would be that `tensorflow` can return `ndarray` from tensors.

**Comment by nhirschey** [comment]
There are probably more elegant ways, but like this?
```fsharp
let t1 = dsharp.tensor [ 0.0 .. 0.2 .. 1.0 ]
let t2 = dsharp.tensor [[1,2],[3,4]]

let a = Array.init t1.shape.[0] (fun i -> t1.[i].toDouble()) 
let b = Array2D.init t2.shape.[0] t2.shape.[1] (fun i j -> t2.[i,j].toDouble())

```

**Comment by nhirschey** [comment]
I was just looking, is this a bug? Is toArray() supposed to return System.Array or FSharp.Collections.Array?

**Comment by dsyme** [comment]
It returns a System.Array, which can then be downcast to the right kind of array 

The exact type of the array is the natural .NET corresponding type for the element type of tensor, seen here for example: https://github.com/DiffSharp/DiffSharp/blob/c1d7eb613539cae3ae11a7cca4df230503ef84e0/src/DiffSharp.Backends.Torch/Torch.RawTensor.fs#L222

If we add `.toArray1D<float>()`, `toArray2D<int64>()` etc. I guess they will do the job of casting to the requested element type before delivering the results.

**Comment by nhirschey** [comment]
This seems to work, though I am sure there is a better way. I do not know how strict the type casting should be. Should `Array1D<float>()` convert any dtype to float or should it only convert certain dtypes? Below converts everything. Pattern matches on dtype could restrict the conversions.

```fsharp
#r "nuget: DiffSharp-lite, 1.0.0-preview-1001211231"

open DiffSharp

dsharp.config(dtype=Dtype.Float32,device=Device.CPU,backend=Backend.Reference)

type Tensor with
    member t.toArray1D<'T>() =
        match t.shape with
        | [| d0 |] -> 
            match box Unchecked.defaultof<'T> with
            | :? bool -> box (Array.init<bool> d0 (fun i -> t.[i].toBool()))
            | :? byte -> box (Array.init<byte> d0 (fun i -> t.[i].toByte()))
            | :? sbyte -> box (Array.init<sbyte> d0 (fun i -> t.[i].toSByte()))
            | :? int16 -> box (Array.init<int16> d0 (fun i -> t.[i].toInt16()))
            | :? int32 -> box (Array.init<int32> d0 (fun i -> t.[i].toInt32()))
            | :? int64 -> box (Array.init<int64> d0 (fun i -> t.[i].toInt64()))
            | :? float32 -> box (Array.init<float32> d0 (fun i -> t.[i].toSingle())) 
            | :? double -> box (Array.init<double> d0 (fun i -> t.[i].toDouble()))
            | _ -> failwith "unknown type"
        | _ -> failwith "expects 1D data"
        :?> 'T[]
    member t.toArray2D<'T>() =
        match t.shape with
        | [| d0; d1 |] -> 
            match box Unchecked.defaultof<'T> with
            | :? bool -> box (Array2D.init<bool> d0 d1 (fun i j -> t.[i,j].toBool()))
            | :? byte -> box (Array2D.init<byte> d0 d1 (fun i j-> t.[i,j].toByte()))
            | :? sbyte -> box (Array2D.init<sbyte> d0 d1 (fun i j-> t.[i,j].toSByte()))
            | :? int16 -> box (Array2D.init<int16> d0 d1 (fun i j-> t.[i,j].toInt16()))
            | :? int32 -> box (Array2D.init<int32> d0 d1 (fun i j-> t.[i,j].toInt32()))
            | :? int64 -> box (Array2D.init<int64> d0 d1 (fun i j-> t.[i,j].toInt64()))
            | :? float32 -> box (Array2D.init<float32> d0 d1 (fun i j-> t.[i,j].toSingle())) 
            | :? double -> box (Array2D.init<double> d0 d1 (fun i j -> t.[i,j].toDouble()))
            | _ -> failwith "unknown type"
        | _ -> failwith "expects 2D data"
        :?> 'T[,]

let t1 = dsharp.tensor [ 0.0 .. 0.2 .. 1.0 ]

t1.toArray1D<single>()
(*
  val it : single [] =
  [|0.0f; 0.200000003f; 0.400000006f; 0.6000000238f; 0.8000000119f; 1.0f|]
*)

t1.toArray1D<float>()
(*
val it : float [] =
  [|0.0; 0.200000003; 0.400000006; 0.6000000238; 0.8000000119; 1.0|]
*)

let t2 = dsharp.stack [t1; t1]
t2.toArray2D<single>()
(*
val it : single [,] =
  [[0.0f; 0.200000003f; 0.400000006f; 0.6000000238f; 0.8000000119f; 1.0f]
   [0.0f; 0.200000003f; 0.400000006f; 0.6000000238f; 0.8000000119f; 1.0f]]
*)
t2.toArray2D<float>()
(*
val it : float [,] =
  [[0.0; 0.200000003; 0.400000006; 0.6000000238; 0.8000000119; 1.0]
   [0.0; 0.200000003; 0.400000006; 0.6000000238; 0.8000000119; 1.0]]
*)

t2.toArray2D<int64>()
(*
val it : int64 [,] = [[0L; 0L; 0L; 1L; 1L; 1L]
                      [0L; 0L; 0L; 1L; 1L; 1L]]
*)

```

**Comment by gbaydin** [comment]
`Tensor.toArray()` returns a [System.Array](https://docs.microsoft.com/en-us/dotnet/api/system.array?view=net-5.0) instance as noted. This supports multi-dimensional arrays up to 32 dimensions (a dotnet limitation).

The most straightforward thing to do is to cast to F# arrays if needed. For example:

```fsharp
// Tensor
t1 = dsharp.tensor [ 0.0 .. 0.2 .. 1.0 ]

// System.Array
let a1 = t1.toArray()

// []<float32>
let a1b = t1.toArray() :?> float32[]

// Tensor
let t2 = dsharp.randn([3;3;3])

// [,,]<float32>
let a2 = t2.toArray() :?> float32[,,]
```

These are already supported. 

We can also add `.toArray1D` etc. to help users.

**Comment by gbaydin** [comment]
I started working on this in #360

**Comment by gbaydin** [comment]
Implemented in #360
